src/algorithm/convexhull.rs

```diff
@@ -0,0 +1,199 @@
+    let mut max_x = Float::min_value();
+    let to_retain: BTreeSet<usize> = (0..points.len() - 1).collect();
+    let mut to_remove: BTreeSet<usize> = BTreeSet::new();
+    for (idx, point) in points.iter().enumerate() {
+        if point.x() < min_x {
+            min_x = point.x();
+            min_x_idx = idx;
+        }
+        if point.x() > max_x {
+            max_x = point.x();
+            max_x_idx = idx;
+        }
+    }
+    to_remove.insert(min_x_idx);
+    to_remove.insert(max_x_idx);
+
+    let p_a = points[min_x_idx];
+    let p_b = points[max_x_idx];
+    // min x and max x points are always part of the hull
+    hull.push(p_a);
+    hull.push(p_b);
+    let mut left_set: Vec<Point<T>> = vec![];
+    let mut right_set: Vec<Point<T>> = vec![];
+    // divide remaining points into left and right
+    // this is a bit hairy
+    for point in &to_retain.difference(&to_remove)
+        .map(|&idx| points[idx])
+        .collect::<Vec<Point<T>>>() {
+        if !point_location(&p_a, &p_b, point) {
+            left_set.push(*point);
+        } else {
+            right_set.push(*point);
+        }
+    }
+    hull_set(&p_a, &p_b, &mut right_set, &mut hull);
+    hull_set(&p_b, &p_a, &mut left_set, &mut hull);
+    // close the polygon
+    let final_element = *hull.first().unwrap();
+    hull.push(final_element);
+    hull
+}
+
+// recursively calculate the convex hull of a subset of points
+fn hull_set<T>(p_a: &Point<T>, p_b: &Point<T>, set: &mut Vec<Point<T>>, hull: &mut Vec<Point<T>>)
+    where T: Float
+{
+    let insert_position = hull.iter().position(|r| r == p_b).unwrap();
+    if set.is_empty() {
+        return;
+    }
+    if set.len() == 1 {
+        hull.insert(insert_position, set[0]);
+        set.remove(0);
+        return;
+    }
+    let mut furthest_distance = Float::min_value();
+    let mut furthest_idx = <usize>::min_value();
+    for (idx, point) in set.iter().enumerate() {
+        let current_distance = pseudo_distance(p_a, p_b, point);
+        if current_distance > furthest_distance {
+            furthest_distance = current_distance;
+            furthest_idx = idx
+        }
+    }
+    // move Point at furthest_point from set into hull
+    let furthest_point = set[furthest_idx];
+    hull.insert(insert_position, set[furthest_idx]);
+    set.remove(furthest_idx);
+    // Determine points to the left of A, furthest_point
+    let mut left_ap: Vec<Point<T>> = vec![];
+    let mut left_pb: Vec<Point<T>> = vec![];
+    for point in set.iter() {
+        if point_location(p_a, &furthest_point, point) {
+            left_ap.push(*point);
+        }
+    }
+    // Determine points to the left of furthest_point, B
+    for point in set.iter() {
+        if point_location(&furthest_point, p_b, point) {
+            left_pb.push(*point);
+        }
+    }
+    // recur
+    hull_set(p_a, &furthest_point, &mut left_ap, hull);
+    hull_set(&furthest_point, p_b, &mut left_pb, hull);
```

Comment:
Yep, I've been messing about with Rayon to try to parallelise this. It'll be in a follow-up, if only for discussion, because we're not doing it anywhere yet…